File: JAVA/24 - Swap Nodes in Pairs.java
/*
 24. Swap Nodes in Pairs (M)

 Approach 1 : Iteration
 Time: O(n)
 Space: O(1)

 Approach 2 : Recursion
 Time: O(n)
 Space:O(n)

*/

/* Approach 1 */
class Solution {
    public ListNode swapPairs(ListNode head) {
        ListNode dummy = new ListNode(0);
        ListNode prev = dummy;
        while (head != null && head.next != null) {
            ListNode tmp = head.next.next;
            prev.next = head.next;
            head.next.next = head;
            head.next = null;
            prev = head;
            head = tmp;
        }
        if (head != null) prev.next = head;
        return dummy.next;
    }
}


/* Approach 2 */
class Solution {
    public ListNode swapPairs(ListNode head) {
        if (head == null || head.next == null) return head;
        ListNode tmp = head.next;
        head.next = swapPairs(head.next.next);
        tmp.next = head;
        return tmp;
    }
}

File: C++/24-Swap_Nodes_in_Pairs.cpp
/*
 24. Swap Nodes in Pairs (Medium)

 Notes:
    Prenode method - O(n)
    Make a dummy pointer points to head, while head->next and head->next->next not nullptr, do:
    1. Make a backup pointer points to head->next->next
    2. Do swap process
    3. Move head pointer to backup pointer
 */

class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        
        // Recursion
        if (!head || !head->next) return head;
        ListNode* next = head->next;
        head->next = swapPairs(head->next->next);
        next->next = head;
        return next;
        
        // Iteration
        /*
        if (!head) return nullptr;
        ListNode* dummy = new ListNode(0);
        ListNode* cur = head;
        ListNode* prev = dummy;
        while (cur && cur->next)
        {
            ListNode* next = cur->next->next;
            prev->next = cur->next;
            prev->next->next = cur;
            prev = cur;
            cur->next = nullptr;
            cur = next;
        }
        if (cur) prev->next = cur;
        return dummy->next;
        */
    }
};